Implement {cli} to build better messages/warning/errors + use testthat3e and snapshot tests

njtierney opened the pull request:
Resolves #418

njtierney commented:
This will also resolve #378

njtierney commented:
Note that the first sentence in `format_error` is the header and subsequent vectors are new lines.

``` r
function_name <- "binomial"
msg <- cli::format_error(
  # first sentence is the header in bold
  c("Wrong function name provided in another model",
    # subsequent sentences are new lines
    "It looks like you're using {.pkg greta}'s {.fun {function_name}} function in \\
    the family argument of another model.",
    "Maybe you want to use {.code family = stats::{function_name}},instead?"
  )
)
stop(msg, call. = FALSE)
#> Error: Wrong function name provided in another model
#> It looks like you're using greta's `binomial()` function in the family argument
#> of another model.
#> Maybe you want to use `family = stats::binomial`,instead?
```

<sup>Created on 2021-06-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.1.0 (2021-05-18)
#>  os       macOS Big Sur 10.16         
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  ctype    en_AU.UTF-8                 
#>  tz       Australia/Perth             
#>  date     2021-06-07                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date       lib source        
#>  backports     1.2.1   2020-12-09 [1] CRAN (R 4.1.0)
#>  cli           2.5.0   2021-04-26 [1] CRAN (R 4.1.0)
#>  crayon        1.4.1   2021-02-08 [1] CRAN (R 4.1.0)
#>  digest        0.6.27  2020-10-24 [1] CRAN (R 4.1.0)
#>  ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.1.0)
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 4.1.0)
#>  fansi         0.5.0   2021-05-25 [1] CRAN (R 4.1.0)
#>  fs            1.5.0   2020-07-31 [1] CRAN (R 4.1.0)
#>  glue          1.4.2   2020-08-27 [1] CRAN (R 4.1.0)
#>  highr         0.9     2021-04-16 [1] CRAN (R 4.1.0)
#>  htmltools     0.5.1.1 2021-01-22 [1] CRAN (R 4.1.0)
#>  knitr         1.33    2021-04-24 [1] CRAN (R 4.1.0)
#>  lifecycle     1.0.0   2021-02-15 [1] CRAN (R 4.1.0)
#>  magrittr      2.0.1   2020-11-17 [1] CRAN (R 4.1.0)
#>  pillar        1.6.1   2021-05-16 [1] CRAN (R 4.1.0)
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.1.0)
#>  purrr         0.3.4   2020-04-17 [1] CRAN (R 4.1.0)
#>  reprex        2.0.0   2021-04-02 [1] CRAN (R 4.1.0)
#>  rlang         0.4.11  2021-04-30 [1] CRAN (R 4.1.0)
#>  rmarkdown     2.8     2021-05-07 [1] CRAN (R 4.1.0)
#>  rstudioapi    0.13    2020-11-12 [1] CRAN (R 4.1.0)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.1.0)
#>  stringi       1.6.2   2021-05-17 [1] CRAN (R 4.1.0)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.1.0)
#>  styler        1.4.1   2021-03-30 [1] CRAN (R 4.1.0)
#>  tibble        3.1.2   2021-05-16 [1] CRAN (R 4.1.0)
#>  utf8          1.2.1   2021-03-12 [1] CRAN (R 4.1.0)
#>  vctrs         0.3.8   2021-04-29 [1] CRAN (R 4.1.0)
#>  withr         2.4.2   2021-04-18 [1] CRAN (R 4.1.0)
#>  xfun          0.23    2021-05-15 [1] CRAN (R 4.1.0)
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 4.1.0)
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.1/Resources/library
```

</details>

![image](https://user-images.githubusercontent.com/6488485/120968362-5d0e2b00-c79b-11eb-8dc2-9e236eb68bec.png)

This is now starting to cover #378, just wort noting.

It will prepare things for #375 nicely, but I think that deserves it's own branch, due to the special error classes. #375 could also serve as an additional time to review the error messages created in this branch (#425)

njtierney commented:
I'm unsure how to handle the `ngettext` function calls which are sometimes used when building an error message

njtierney commented:
Due to the implementation of {cli}, the decision to use a pluralised sentence/word it needs to have the thing that is plural before the pluralisation decision point.

What this means is that if we want to use {cli} to build the sentence we need to have:

> running 1 sampler in parallel, each on **2 cores or fewer**

Compared to:

> running 2 samplers in parallel, each **on up to 2 cores**

@goldingn do you have a preference for the sentence?

Here are the two implementations, the {cli} one is a bit more compact/concise and uses interpolated strings.

``` r
# old print

print_sample_core_info <- function(n_cores,
                                    samplers){
    cores_text <- ifelse(n_cores == 1,
                         "1 core",
                         sprintf("up to %i cores", n_cores)
    )
    
    msg <- sprintf(
      "\nrunning %i samplers in parallel, each on %s\n\n",
      length(samplers),
      cores_text
    )
    cli::cli_alert_info(msg)
}

print_sample_core_info(n_cores = 1, samplers = 1)
#> ℹ running 1 samplers in parallel, each on 1 core
print_sample_core_info(n_cores = 1, samplers = 1:2)
#> ℹ running 2 samplers in parallel, each on 1 core
print_sample_core_info(n_cores = 2, samplers = 1)
#> ℹ running 1 samplers in parallel, each on up to 2 cores
print_sample_core_info(n_cores = 2, samplers = 1:2)
#> ℹ running 2 samplers in parallel, each on up to 2 cores

# new method
print_sample_core_info2 <- function(n_cores = 1,
                                   samplers = 1:2){
    msg <- cli::format_message(
      "running {length(samplers)} sampler{?s} in parallel, each on \\
      {n_cores} core{?/s or fewer}",
    )
    cli::cli_alert_info(msg)
}

print_sample_core_info2(n_cores = 1, samplers = 1)
#> ℹ running 1 sampler in parallel, each on 1 core
print_sample_core_info2(n_cores = 1, samplers = 1:2)
#> ℹ running 2 samplers in parallel, each on 1 core
print_sample_core_info2(n_cores = 2, samplers = 1)
#> ℹ running 1 sampler in parallel, each on 2 cores or fewer
print_sample_core_info2(n_cores = 2, samplers = 1:2)
#> ℹ running 2 samplers in parallel, each on 2 cores or fewer
```

<sup>Created on 2021-06-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.1.0 (2021-05-18)
#>  os       macOS Big Sur 10.16         
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  ctype    en_AU.UTF-8                 
#>  tz       Australia/Perth             
#>  date     2021-06-08                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date       lib source        
#>  backports     1.2.1   2020-12-09 [1] CRAN (R 4.1.0)
#>  cli           2.5.0   2021-04-26 [1] CRAN (R 4.1.0)
#>  crayon        1.4.1   2021-02-08 [1] CRAN (R 4.1.0)
#>  digest        0.6.27  2020-10-24 [1] CRAN (R 4.1.0)
#>  ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.1.0)
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 4.1.0)
#>  fansi         0.5.0   2021-05-25 [1] CRAN (R 4.1.0)
#>  fs            1.5.0   2020-07-31 [1] CRAN (R 4.1.0)
#>  glue          1.4.2   2020-08-27 [1] CRAN (R 4.1.0)
#>  highr         0.9     2021-04-16 [1] CRAN (R 4.1.0)
#>  htmltools     0.5.1.1 2021-01-22 [1] CRAN (R 4.1.0)
#>  knitr         1.33    2021-04-24 [1] CRAN (R 4.1.0)
#>  lifecycle     1.0.0   2021-02-15 [1] CRAN (R 4.1.0)
#>  magrittr      2.0.1   2020-11-17 [1] CRAN (R 4.1.0)
#>  pillar        1.6.1   2021-05-16 [1] CRAN (R 4.1.0)
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.1.0)
#>  purrr         0.3.4   2020-04-17 [1] CRAN (R 4.1.0)
#>  reprex        2.0.0   2021-04-02 [1] CRAN (R 4.1.0)
#>  rlang         0.4.11  2021-04-30 [1] CRAN (R 4.1.0)
#>  rmarkdown     2.8     2021-05-07 [1] CRAN (R 4.1.0)
#>  rstudioapi    0.13    2020-11-12 [1] CRAN (R 4.1.0)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.1.0)
#>  stringi       1.6.2   2021-05-17 [1] CRAN (R 4.1.0)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.1.0)
#>  styler        1.4.1   2021-03-30 [1] CRAN (R 4.1.0)
#>  tibble        3.1.2   2021-05-16 [1] CRAN (R 4.1.0)
#>  utf8          1.2.1   2021-03-12 [1] CRAN (R 4.1.0)
#>  vctrs         0.3.8   2021-04-29 [1] CRAN (R 4.1.0)
#>  withr         2.4.2   2021-04-18 [1] CRAN (R 4.1.0)
#>  xfun          0.23    2021-05-15 [1] CRAN (R 4.1.0)
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 4.1.0)
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.1/Resources/library
```

</details>

goldingn commented:
Message formatting aside, does it make sense to report this using cli alerts? This bit isn't a CLI, as everything only affects objects within an R session rather than affecting the system. I think having messages formatted as cli alerts will confuse people.

Is this formatting restriction the same in glue? Or when using ngettext()?

njtierney commented:
> Message formatting aside, does it make sense to report this using cli alerts? This bit isn't a CLI, as everything only affects objects within an R session rather than affecting the system. I think having messages formatted as cli alerts will confuse people.

My understanding was that the cli alert was giving nice icons (like the tick, cross, "i" symbol for warning or general alert), combined with glue string interpolation, pluralisation, and other niceties, like specially formatting error messages, providing classes for data objects that are functions, classes, and turning vectors into lists - e.g.,

``` r
vec <- c("pkg1", "pkg2", "pkg3")
cli::cli_alert_info("The following packages were installed: {vec}")
#> ℹ The following packages were installed: pkg1, pkg2, and pkg3
```

I'm not sure using cli will make things confusing, but it could make them appear inconsistent in terms of text and styling if mixed with other messages that don't have similar formatting. In this case I would consider it a helpful alert that would generally appear as white text - "hey, here's some info about the samplers/cores/chains".  

That said, the progress bar would be generally rendered in orangeish text, and the message above would be in white, so it might look a bit strange?

> Is this formatting restriction the same in glue? Or when using ngettext()?

Glue only does string interpolation, and does not do pluralisation, that is only done in `cli`

I think I _completely_ misunderstood `ngettext` 😅  - I read "translation" as allowing the text to get translated to other languages (french, spanish, chinese, etc):

> If Native Language Support (NLS) was enabled in this build of R (see the bindtextdomain() example), attempt to translate character vectors or set where the translations are to be found.

But reading it now it specifically helps with pluralisation. This can be implemented with `cli`

njtierney commented:
This PR is now ready for a review (of error messages).

I've given the error messages a clean up and clarified where I can, but I still think that there is scope for a further review of the error messages, perhaps to be in line with suggestions on https://style.tidyverse.org/error-messages.html

This could be implemented in another PR, or we can spend some time reviewing the messages now and then also close #382 .

A lot of the tests for greta will now fail, as they are checking the error/warning messages, which have changed. Fixing these could be tedious, if it is quite tedious, we could change the tests over to use snapshot tests, in Testthat 3e. However, that would involve also addressing issues with using mock (#381). So things are a bit interconnected at the moment.

njtierney commented:
So thinking about this in full, this should resolve the following issues:

- #418
- #382 
- #381

njtierney commented on tests/testthat/_snaps/as_data.md:
```diff
@@ -0,0 +1,106 @@
+      as_data(NULL)
+    Error <simpleError>
+      Object cannot be coerced to <greta_array>
+      Objects of class <NULL> cannot be coerced to a <greta_array>
```
An example comment @goldingn

goldingn commented on tests/testthat/_snaps/misc.md:
```diff
@@ -0,0 +1,159 @@
+    Code
+      model(bernoulli(0.5))
+    Error <simpleError>
+      model contains a discrete random variable that doesn't have a fixed
```
we don't know that the user wants to do MCMC, so reword to:
```
model contains a discrete random variable that doesn't have a fixed value, so inference cannot be carried out
```

goldingn commented on tests/testthat/_snaps/operators.md:
```diff
@@ -0,0 +1,14 @@
+    Code
+      a %*% c
+    Error <simpleError>
+      cannot coerce type 'builtin' to vector of type 'character'
```
not so informative, but maybe this is a silly test? Or is it an error in the error-message code?

njtierney replied:
I'm actually not sure what is happening here, might need to take a closer look.

njtierney replied:
Passed `stop` `dim` instead of the `msg` !

goldingn commented on tests/testthat/_snaps/simulate.md:
```diff
@@ -0,0 +1,39 @@
+    Code
+      simulate(m, nsim = 0)
+    Error <simpleError>
+      value must be a positive integer
```
should say `nsim`?

njtierney replied:
This is that issue with the argument `nsim` not being passed down to the fundamental checking function. Not sure how to address this.

goldingn commented on tests/testthat/_snaps/syntax.md:
```diff
@@ -0,0 +1,73 @@
+    Code
+      distribution(y)
+    Error <simpleError>
+      `distribution()` expects object of type, <greta_array>
```
rogue comma

goldingn replied:
good band name that ^

goldingn commented on tests/testthat/_snaps/variables.md:
```diff
@@ -0,0 +1,86 @@
+    Code
+      variable(upper = NA)
+    Error <simpleError>
+      lower and upper must be numeric
```
too verbose? should only report on upper?

njtierney replied:
The error message I think captures both cases, where lower can be numeric and upper something else?

goldingn replied:
could split the test in two? check each one separately and then the conflict between them? Though that's probably yak shaving. fine to leave it as is

njtierney replied:
I've added a couple more tests for this - I think that covers it.

goldingn commented on tests/testthat/_snaps/variables.md:
```diff
@@ -0,0 +1,86 @@
+      cholesky_variable(dim = rep(2, 3))
+    Error <simpleError>
+      dim can either be a scalar or a vector of length 2
+      However its dimension is: 2, 2, and 2
```
shouldn't say 'its dimension', because the 'it' here _is_ the dimension

njtierney replied:
I'm not sure how to rephrase this - would something like this work in this instance?

```
dim can either be a scalar or a vector of length 2
but dim has length of 3: 2, 2, 2
```

goldingn replied:
yeah, that's perfect

njtierney commented on tests/testthat/_snaps/as_data.md:
```diff
@@ -0,0 +1,106 @@
+      <greta_array> must contain the same type
+      Cannot coerce <matrix> to a <greta_array> unless all columns are <numeric>,
+      <integer> or <logical>. This <matrix> had type:
+      <character>
```
Done!

njtierney commented on tests/testthat/_snaps/calculate.md:
```diff
@@ -0,0 +1,137 @@
+      The following object passed to `calculate()` is not a <greta array>:
+      "x"
+      Perhaps you forgot to explicitly name other arguments?
+      Perhaps you forgot to explicitly name other arguments?
```
I couldn't work this one out - I think it printed for each object, will need to mull this over.

njtierney replied:
Figured it out - `stop` was adding `hint` as an argument.

njtierney commented on tests/testthat/_snaps/calculate.md:
```diff
@@ -0,0 +1,137 @@
+    Code
+      calc_x <- calculate(x, nsim = 0)
+    Error <simpleError>
+      value must be a positive integer
```
Yup - that's because the error message is called from `check_positive_integer()`. I'm not sure how to get the argument out here, it might involve some NSE magic?

goldingn replied:
I think that's what the `name` argument to `check_positive_integer()` used to do, you should be able to pass that through like this: `check_positive_integer(nsim, "nsim")` and make sure it's plumbed into the cli code. Though an NSE solution would be neat :)

njtierney replied:
Ah that is neat! And avoids NSE magicks 🧙

njtierney commented on tests/testthat/_snaps/extract_replace_combine.md:
```diff
@@ -0,0 +1,101 @@
+    Code
+      abind(a, c, along = 5)
+    Error <simpleError>
+      invalid 'enclos' argument of type 'character'
```
Yeah I've got no idea. My guess would be something about `c` being a function?

njtierney replied:
Ah it was a combination of two things

First, there was a control statement that was:

```
if (!along %in% 0:n)
```

Instead of

```
if (!(along %in% 0:n))
```

Since `along` is 5 - and !5 is FALSE, this didn't make any sense...

I also forgot to pass a vector to format_error - now fixed:

```
if (!(along %in% 0:n)) {
    msg <- cli::format_error(
      c(
        "{.arg along} must be between 0 and {n}",
        "Instead {.arg along} was {.val {along}}"
      )
    )
```

njtierney commented on tests/testthat/_snaps/calculate.md:
```diff
@@ -0,0 +1,137 @@
+    Code
+      calc_a <- calculate(a, y, values = draws)
+    Error <simpleError>
+      `nsim` must be set to sample <greta_array> with distributions not in
```
I've added that line in - did you just want that first sentence replaced? The updated one should appear now

goldingn replied:
yeah, I was thinking just the main error message to that, leave the further context as-is

njtierney commented on tests/testthat/_snaps/inference.md:
```diff
@@ -0,0 +1,405 @@
+      mcmc(m, chains = 1, n_samples = 1, warmup = 0, initial_values = initials(z = 1e+20))
+    Error <simpleError>
+      The log density could not be evaluated at these initial values
+      Try using `calculate()` to see whether they lead to values of other
```
done

njtierney commented on tests/testthat/_snaps/inference.md:
```diff
@@ -0,0 +1,405 @@
+    Code
+      model(a, b)
+    Error <simpleError>
+      `model()` arguments must be <greta_array>
```
done
